### PR TITLE
Implement async generators.

### PR DESCRIPTION
**Pull request by mnieper**
Implement async generators.

Handle throw in for-on bodies.

Allow a yield expression after await.

----

Please review (and contribute more tests if you like :-)).

### CODE REVIEW

**Comment by domenic** [comment]
I couldn't find the Observable implementation in here? Which async generator proposal are you implementing?

**Comment by mnieper** [comment]
I am implementing https://github.com/jhusain/asyncgenerator. As far as I understand, the core of the proposal is a new protocol (the observable and the observer interface) and two syntactic additions, async generators and for-on loops that conform to the protocol. Both have been implemented.

According to https://github.com/jhusain/asyncgenerator#observable-composition, the various functions for observables are not proposed for standardization but left to library writers. Thus they are not included in this PR.

Any observable can be easily extended by adding methods in the prototype chain of the `prototype` property of any async generator function.

**Comment by domenic** [comment]
I think I understand, thanks.

File: snippets.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+
+/*
+    return createFunctionBody(parseStatements `
```
Comment (arv) [inline snippets.js]: Remove from PR?

File: src/codegeneration/AsyncGeneratorTransformer.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  // don't transform children functions
+  transformFunctionDeclaration(tree) {return tree;}
+  transformFunctionExpression(tree) {return tree;}
+  transformSetAccessor(tree) {return tree;}
+  transformGetAccessor(tree) {return tree;}
+  transformPropertyMethodAssignment(tree) {return tree;}
+  transformArrowFunctionExpression(tree) {return tree;}
+
+   /**
+   *
```
Comment (arv) [inline src/codegeneration/AsyncGeneratorTransformer.js]: remove "empty" line

File: src/codegeneration/ForOnTransformer.js [hunk 3]
```diff
@@ -40,12 +40,16 @@ import {
  *   var $result = undefined;
  *   let i;
  *   await $traceurRuntime.observableForEach(o, async function ($value) {
- *     $observer = this;
- *     i = $value;
- *     if (continue) return;
- *     if (break) { $result = 0; $observer.return(); return; }
- *     if (return) { $result = {v: undefined}; $observer.return(); return; }
- *   });
+.*...  try {
```
Comment (arv) [inline src/codegeneration/ForOnTransformer.js]: I'm not sure what the dots are for?

File: src/outputgeneration/ParseTreeWriter.js [hunk 4]
```diff
@@ -712,9 +714,15 @@ export class ParseTreeWriter extends ParseTreeVisitor {
 
   visitFunction_(tree) {
     this.writeAnnotations_(tree.annotations);
+    if (tree.isAsyncGenerator()) {
+      this.write_(new IdentifierToken(tree.functionKind.location, ASYNC));
```
Comment (arv) [inline src/outputgeneration/ParseTreeWriter.js]: You should be able to do `write_(ASYNC)`

File: src/outputgeneration/ParseTreeWriter.js [hunk 5]
```diff
@@ -712,9 +714,15 @@ export class ParseTreeWriter extends ParseTreeVisitor {
...
+    }
     if (tree.isAsyncFunction())
       this.write_(tree.functionKind);
     this.write_(FUNCTION);
+    if (tree.isAsyncGenerator()) {
+      this.write_(new Token(STAR, null));
```
Comment (arv) [inline src/outputgeneration/ParseTreeWriter.js]: same

File: src/runtime/async.js [hunk 6]
```diff
@@ -12,6 +12,127 @@
...
+            return;
+          }
+          if (result.done) {
+            ctx.done = true;
+          }
+          return result;
+        });
+  }
+}
+    // und XXX teste ob = void 0 wegkann
```
Comment (arv) [inline src/runtime/async.js]: remove

File: src/syntax/PredefinedName.js [hunk 7]
```diff
@@ -19,6 +19,7 @@ export var ARGUMENTS = 'arguments';
 export var ARRAY = 'Array';
 export var AS = 'as';
 export var ASYNC = 'async';
+export var ASYNC_STAR = 'asyncStar'; // internal kind for async generators
```
Comment (arv) [inline src/syntax/PredefinedName.js]: maybe use `'async star'` or something else that isn't a valid identifier.

File: test/feature/AsyncGenerators/ThisAndArguments.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+// Async.
+
+async function* f(a) {
+  yield this[0];
+  yield arguments[0];
+}
+
+(async function() {
+  var list = [];
+  var g = f.apply([1], [2]);
```
Comment (arv) [inline test/feature/AsyncGenerators/ThisAndArguments.js]: var g = f.call([1], 2);

is simpler

File: test/feature/AsyncGenerators/AsyncGenerator.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+// Async.
+
+async function* f() {
+  yield 1;
+}
+
+(async function() {
+  var list = [];
+  var g = f();
+  for (i on g) {
```
Comment (arv) [inline test/feature/AsyncGenerators/AsyncGenerator.js]: missing var for i here and in the other tests.

File: test/feature/AsyncGenerators/ThrowInAsyncGenerator.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  var g = f();
+  try {
+    for (i on g) {
+      if (i == 2) {
+        throw 'ex';
+      }
+      list.push(i);
+    }
+  } catch (ex) {
+    assert.deepEqual(ex, 'ex');
```
Comment (arv) [inline test/feature/AsyncGenerators/ThrowInAsyncGenerator.js]: assert.equal

**Comment by arv** [comment]
Only a few nits.

File: src/codegeneration/ForOnTransformer.js (hunk 3 above)
Comment (mnieper) [inline src/codegeneration/ForOnTransformer.js]: Blame gedit...

**Comment by mnieper** [comment]
Adressed the issues and also removed an unnecessary try-catch in the generated code.

File: src/outputgeneration/ParseTreeWriter.js [hunk 11]
```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import {Token} from '../syntax/Token.js';
```
Comment (arv) [inline src/outputgeneration/ParseTreeWriter.js]: Useless imports

There is a tool that will warn about these:

```sh
$ git diff master --name-only | grep src/ | xargs node tools/find-unused-imports.js 
src/codegeneration/AsyncGeneratorTransformer.js:21:3: Unused import: createIdentifierToken
src/codegeneration/AsyncGeneratorTransformer.js:30:3: Unused import: parseStatements
src/codegeneration/AsyncGeneratorTransformer.js:38:3: Unused import: MemberExpression
src/codegeneration/AsyncGeneratorTransformer.js:42:9: Unused import: VARIABLE_DECLARATION_LIST
src/codegeneration/ForOnTransformer.js:22:33: Unused import: id
src/codegeneration/ForOnTransformer.js:23:3: Unused import: createVariableStatement
src/codegeneration/InnerForOnTransformer.js:24:3: Unused import: BreakStatement
src/codegeneration/InnerForOnTransformer.js:30:3: Unused import: createArgumentList
src/codegeneration/InnerForOnTransformer.js:32:3: Unused import: createAssignmentExpression
src/outputgeneration/ParseTreeWriter.js:15:9: Unused import: Token
src/outputgeneration/ParseTreeWriter.js:16:9: Unused import: IdentifierToken
src/syntax/trees/ParseTree.js:15:13: Unused import: ParseTreeType
```

File: src/runtime/async.js [hunk 12]
```diff
@@ -22,11 +143,51 @@ function observeForEach(observe, next) {
...
+
+class DecoratedGenerator {
+  constructor(_generator, _onDone) {
+    this[generator] = _generator;
+    this[onDone] = _onDone;
+  }
+
+  next(value) {
+    var result = this[generator].next(value);
+    if (result !== void 0 && result.done) {
```
Comment (arv) [inline src/runtime/async.js]: We haven't been very strict about `void 0` vs `undefined` but please be consistent. Previously, in this file you used `undefined`.

**Comment by arv** [comment]
LGTM. A few nits

File: src/codegeneration/AsyncGeneratorTransformer.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+  // don't transform children functions
+  transformFunctionDeclaration(tree) {return tree;}
+  transformFunctionExpression(tree) {return tree;}
+  transformSetAccessor(tree) {return tree;}
+  transformGetAccessor(tree) {return tree;}
+  transformPropertyMethodAssignment(tree) {return tree;}
+  transformArrowFunctionExpression(tree) {return tree;}
+
+   /**
+   * @param {FunctionBody} tree
```
Comment (arv) [inline src/codegeneration/AsyncGeneratorTransformer.js]: Incorrect formatting:

```js
/**
 * ...
 */
```

File: src/outputgeneration/ParseTreeWriter.js (hunk 11 above)
Comment (mnieper) [inline src/outputgeneration/ParseTreeWriter.js]: The last warning is erroneous.

**Comment by mnieper** [comment]
Thanks for reviewing! Pushed the corrections.

**Comment by arv** [comment]
I noticed that our tests are broken so I'm holding off this one until that has been resolved.

**Comment by mnieper** [comment]
No problem. I'll check again tomorrow (it's now 10 pm here).

Once this is accepted, we can start adding documentation and maybe a demo.

**Comment by arv** [comment]
Committed as c8815199e74551f064ff5bb5b5fa122df8e5cbf3

**Pull request closed without merging**